library/kani/src/futures.rs

```diff
@@ -41,3 +44,216 @@ const NOOP_RAW_WAKER: RawWaker = {
+#[derive(Default)]
+pub struct NondeterministicScheduling;
+
+impl SchedulingStrategy for NondeterministicScheduling {
+    #[cfg(kani)]
+    fn pick_task(&mut self, num_tasks: usize) -> (usize, bool) {
+        let index = crate::any();
+        crate::assume(index < num_tasks);
+        (index, true)
+    }
+
+    #[cfg(not(kani))]
+    fn pick_task(&mut self, _num_tasks: usize) -> (usize, bool) {
+        panic!("Nondeterministic scheduling is only available when running Kani.")
+    }
+}
+
+/// A restricted form of nondeterministic scheduling to have some fairness.
+///
+/// Each task has a counter that is increased when it is scheduled.
+/// If a task has reached a provided limit, it cannot be scheduled anymore until all other tasks have reached the limit too,
+/// at which point all the counters are reset to zero.
+pub struct NondetFairScheduling {
+    counters: [u8; MAX_TASKS],
+    limit: u8,
+}
+
+impl NondetFairScheduling {
+    #[inline]
+    pub fn new(limit: u8) -> Self {
+        Self { counters: [limit; MAX_TASKS], limit }
+    }
+}
+
+impl SchedulingPlan for NondetFairScheduling {
+    #[cfg(kani)]
+    fn pick_task(&mut self, num_tasks: usize) -> (usize, bool) {
+        if self.counters[0..num_tasks] == [0; MAX_TASKS][0..num_tasks] {
+            self.counters = [self.limit; MAX_TASKS];
+        }
+        let index = kani::any();
+        kani::assume(index < num_tasks);
+        kani::assume(self.counters[index] > 0);
+        self.counters[index] -= 1;
+        (index, true)
+    }
+
+    #[cfg(not(kani))]
+    fn pick_task(&mut self, _num_tasks: usize) -> (usize, bool) {
+        panic!("Nondeterministic scheduling is only available when running Kani.")
+    }
+}
+
+pub(crate) struct Scheduler {
+    /// Using a Vec instead of an array makes the runtime jump from 40s to almost 10min if using Vec::with_capacity and leads to out of memory with Vec::new (even with 64 GB RAM).
+    tasks: [Option<BoxFuture>; MAX_TASKS],
+    num_tasks: usize,
+    num_running: usize,
+}
+
+impl Scheduler {
+    /// Creates a scheduler with an empty task list
+    #[inline]
+    pub(crate) const fn new() -> Scheduler {
+        const INIT: Option<BoxFuture> = None;
+        Scheduler { tasks: [INIT; MAX_TASKS], num_tasks: 0, num_running: 0 }
+    }
+
+    /// Adds a future to the scheduler's task list, returning a JoinHandle
+    #[inline] // to work around linking issue
+    pub(crate) fn spawn<F: Future<Output = ()> + Sync + 'static>(&mut self, fut: F) -> JoinHandle {
+        let index = self.num_tasks;
+        self.tasks[index] = Some(Box::pin(fut));
+        self.num_tasks += 1;
+        self.num_running += 1;
+        JoinHandle { index }
+    }
+
+    /// Runs the scheduler with the given scheduling plan until all tasks have completed
+    #[inline] // to work around linking issue
+    fn run(&mut self, mut scheduling_plan: impl SchedulingStrategy) {
+        let waker = unsafe { Waker::from_raw(NOOP_RAW_WAKER) };
```

Comment:
what does this mean? No waker?

Reply:
It is a waker, but it does nothing. That's OK because the scheduler does not rely on wakers. If we implemented wakers as in tokio, that would mean having another queue that slows things down.

Reply:
Can we use https://docs.rs/futures/latest/futures/task/fn.noop_waker.html instead?

Reply:
We could, but it would mean pulling in a dependency just for this one type, which isn't very hard to define ourselves. Also, by defining it ourselves, we retain the option of optimizing it for Kani in the future.

Reply:
fair. Thanks